File: dotnet/CucumberExpressions/Generation/CombinatorialGeneratedExpressionFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CucumberExpressions.Generation;

internal class CombinatorialGeneratedExpressionFactory
{
    // 256 generated expressions ought to be enough for anybody
    private const int MAX_EXPRESSIONS = 256;
    private readonly string _expressionTemplate;
    private readonly List<List<IParameterType>> _parameterTypeCombinations;

    public CombinatorialGeneratedExpressionFactory(string expressionTemplate,
            List<List<IParameterType>> parameterTypeCombinations)
    {
        _expressionTemplate = expressionTemplate;
        _parameterTypeCombinations = parameterTypeCombinations;
    }

    public GeneratedExpression[] GenerateExpressions()
    {
        var generatedExpressions = new List<GeneratedExpression>();
        var permutation = new Stack<IParameterType>(_parameterTypeCombinations.Count);
        GeneratePermutations(generatedExpressions, permutation);
        return generatedExpressions.ToArray();
    }

    private void GeneratePermutations(List<GeneratedExpression> generatedExpressions,
            Stack<IParameterType> permutation)
    {
        if (generatedExpressions.Count >= MAX_EXPRESSIONS)
        {
            return;
        }

        if (permutation.Count == _parameterTypeCombinations.Count)
        {
            var permutationCopy = permutation.Reverse().ToArray();
            generatedExpressions.Add(new GeneratedExpression(_expressionTemplate, permutationCopy));
            return;
        }

        var parameterTypes = _parameterTypeCombinations[permutation.Count];
        foreach (var parameterType in parameterTypes)
        {
            // Avoid recursion if no elements can be added.
            if (generatedExpressions.Count >= MAX_EXPRESSIONS)
            {
                return;
            }
            permutation.Push(parameterType);
            GeneratePermutations(generatedExpressions, permutation);
            permutation.Pop();
        }
    }
}

File: python/cucumber_expressions/combinatorial_generated_expression_factory.py
from cucumber_expressions.generated_expression import GeneratedExpression
from cucumber_expressions.parameter_type import ParameterType


class CombinatorialGeneratedExpressionFactory:
    def __init__(self, expression_template, parameter_type_combinations):
        self.expression_template = expression_template
        self.parameter_type_combinations = parameter_type_combinations
        # 256 generated expressions ought to be enough for anybody
        self.MAX_EXPRESSIONS = 256

    def generate_expressions(self) -> list[GeneratedExpression]:
        generated_expressions = []
        self.generate_permutations(generated_expressions, 0, [])
        return generated_expressions

    def generate_permutations(
        self,
        generated_expressions: list[GeneratedExpression],
        depth: int,
        current_parameter_types: list[ParameterType],
    ):
        if len(generated_expressions) >= self.MAX_EXPRESSIONS:
            return
        if depth == len(self.parameter_type_combinations):
            generated_expressions.append(
                GeneratedExpression(self.expression_template, current_parameter_types)
            )
            return
        for i in range(0, len(self.parameter_type_combinations[depth])):
            if len(generated_expressions) >= self.MAX_EXPRESSIONS:
                return
            new_current_parameter_types = current_parameter_types.copy()
            new_current_parameter_types.append(
                self.parameter_type_combinations[depth][i]
            )
            self.generate_permutations(
                generated_expressions, depth + 1, new_current_parameter_types
            )
